### PR TITLE
Cleanup and improve Makefile build logic.

### PR DESCRIPTION
**Pull request by baulig**
We now use custom .sources file for intermediate builds, such as `System/net_4_x_bare_System.dll.sources` and `net_4_x_secxml_System.dll.sources`.

The goal is to get rid of the `SECURITY_DEP` and related conditionals to make it easier to use code from corefx.

### CODE REVIEW

**Comment by baulig** [comment]
This PR is the "stable" version.  https://github.com/mono/mono/pull/4798 adds more experimental stuff.

**Comment by migueldeicaza** [comment]
Being one of the maintainers of the staged build (a duty that fell on me, just to get reliable builds) this commit is something that I do not understand in various ways.

I need more information about this change, the statement is to "get rid of `SECURITY_DEP` and related conditionals to make it easier to reuse code from corefx".

Because the patch does not seem to be related to this, and I am trying to figure out why.

So follow up questions:

* How does this change to the makefile rules eliminate the need for `SECURITY_DEP`?   This define is in place so we can build libraries in stages that incrementally build with the right dependencies the surfaced API that we require.

* How does this clean and improve the build logic?   I do not understand how the commit summary, or the message are related to the patch itself.

**Comment by marek-safar** [comment]
@migueldeicaza I think all what this patch does (module the MAKE changes which should go to separate PR) it only allows intermediate builds .sources customization.

@baulig how are you going to use the intermediate steps customization

File: mcs/build/rules.make [hunk 1]
```diff
@@ -25,6 +25,10 @@ Q=$(if $(V),,@)
 Q_MCS=$(if $(V),,@echo "$(if $(MCS_MODE),MCS,CSC)     [$(intermediate)$(PROFILE)] $(notdir $(@))";)
 Q_AOT=$(if $(V),,@echo "AOT     [$(intermediate)$(PROFILE)] $(notdir $(@))";)
 
+MAKE_Q=$(if $(V),,-s)
+MAKE_V=$(if $(V),V=1,)
```
Comment (luhenry) [inline mcs/build/rules.make]: These are redondant: if `V=1` or `-s` are defined when calling `make`, then it's going to be passed down with `$(MAKE)`

File: mcs/class/System/Makefile [hunk 2]
```diff
@@ -150,27 +150,27 @@ ifeq (bare/,$(intermediate))
 build-bare:
 else
 $(bare_libdir)/System.dll:
-	$(MAKE) intermediate=bare/ $(bare_libdir)/System.dll
+	$(MAKE) $(MAKE_V) intermediate=bare/ $(bare_libdir)/System.dll
```
Comment (luhenry) [inline mcs/class/System/Makefile]: See https://github.com/mono/mono/pull/4794/files#r115024370

File: mcs/class/System/Makefile [hunk 3]
```diff
@@ -150,27 +150,27 @@ ifeq (bare/,$(intermediate))
...
 endif
 
 ifneq (secxml/,$(intermediate))
 $(secxml_libdir)/System.dll: $(bare_libdir)/System.dll $(bare_libdir)/System.Xml.dll $(MONO_SECURITY_DLL) $(bare_libdir)/System.dll
-	$(MAKE) intermediate=secxml/ $(secxml_libdir)/System.dll
+	$(MAKE) $(MAKE_V) intermediate=secxml/ $(secxml_libdir)/System.dll
```
Comment (luhenry) [inline mcs/class/System/Makefile]: See https://github.com/mono/mono/pull/4794/files#r115024370

File: mcs/build/library.make [hunk 4]
```diff
@@ -28,11 +28,23 @@ LIB_MCS_FLAGS += $(patsubst %,-r:%.dll, $(subst =,=$(topdir)/class/lib/$(PROFILE
 
 sourcefile = $(LIBRARY).sources
 
+profile_library = $(PROFILE)_$(LIBRARY)
```
Comment (luhenry) [inline mcs/build/library.make]: Why not simply have `$(PROFILE)_$(LIBRARY_SUBDIR)_$(subst /,-,$(intermediate))_$(LIBRARY)` ? There are going to be some consecutive `_` but that's for generated files so it does not matter.

**Pull request closed without merging**